Gemfile

```diff
@@ -10,3 +10,5 @@ group :development do
   gem "method_profiler"
   gem "ruby-prof"
 end
+
+gem "memoizable"
```

Comment:
I'm unsure about adding a dependency, it doesn't seem necessary. I'm not too keen on the syntax of Memoizable, I feel like `||=` would be just as good and probably perform slightly faster, e.g.

```ruby
def ret_validator
  @ret_validator ||= Contract.make_validator(ret_contract)
end
```

Factoring out these chunks as methods has improved the code though, so whether or not Memoizable is used to do it probably doesn't matter, I'm just edgy about it. On the plus side it looks well maintained and supports the same Rubies we do.